File: unity-integration/Assets/GameClustering/OutboundMessage.cs
using System;
using System.IO;

namespace GameClustering
{
    public class OutboundMessage : IDisposable
    {
        public const int MessageSize = 512;
        
        private readonly MemoryStream _memoryStream;
        private bool _disposed;
        private int _payloadSize;
        public OutboundMessage(){
            _memoryStream = new MemoryStream(new byte[MessageSize]);
        }
        
        public void Ack(bool ack)
        {
            _memoryStream.Position = InboundMessage.AckPos;
            _memoryStream.WriteByte(ack?(byte)1:(byte)0);
        }
        
        public void Type(int type)
        {
            var bytes = BitConverter.GetBytes(type);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            _memoryStream.Position = InboundMessage.TypePos;
            _memoryStream.Write(bytes,0,4);
        }
        
        public void MessageId(int messageId)
        {
            var bytes = BitConverter.GetBytes(messageId);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            _memoryStream.Position = InboundMessage.MessageIdPos;
            _memoryStream.Write(bytes,0,4);
        }
        
        public void ConnectionId(long connectionId)
        {
            var bytes = BitConverter.GetBytes(connectionId);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            _memoryStream.Position = InboundMessage.ConnectionIdPos;
            _memoryStream.Write(bytes,0,8);
        }
        
        public void Sequence(int sequence)
        {
            var bytes = BitConverter.GetBytes(sequence);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            _memoryStream.Position = InboundMessage.SequencePos;
            _memoryStream.Write(bytes,0,4);
        }
        public void Timestamp(long timestamp)
        {
            var bytes = BitConverter.GetBytes(timestamp);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }
            _memoryStream.Position = InboundMessage.TimestampPos;
            _memoryStream.Write(bytes,0,8);
        }
        public void Payload(byte[] payload)
        {
            _memoryStream.Position = InboundMessage.PayloadPos;
            _memoryStream.Write(payload,0,payload.Length);
            _payloadSize = payload.Length;
        }
        public byte[] Message()
        {
            var payload = new byte[InboundMessage.PayloadPos + _payloadSize];
            _memoryStream.Position = InboundMessage.AckPos;
            _memoryStream.Read(payload, 0, payload.Length);
            return payload;
        }
        public void Dispose() => Dispose(true);
        protected virtual void Dispose(bool disposing)
        {
            if (_disposed)
            {
                return;
            }
            if (disposing)
            {
                _memoryStream?.Dispose();
            }
            _disposed = true;
        }
        ~OutboundMessage() => Dispose(false);
    }
}

File: modules/protocol/src/main/java/com/icodesoftware/protocol/OutboundMessage.java
package com.icodesoftware.protocol;

import java.nio.ByteBuffer;

/**
 * Created by yinghu lu on 9/26/2020.
 */
public class OutboundMessage {
    public static int MESSAGE_SIZE = 512;
    private ByteBuffer message;
    private int payloadSize;
    public OutboundMessage(){
        message = ByteBuffer.allocate(MESSAGE_SIZE);
    }
    public void ack(boolean ack){
        message.put(InboundMessage.ACK_POS,ack?(byte)1:(byte)0);
    }
    public void type(int type){
        message.putInt(InboundMessage.TYPE_POS,type);
    }
    public void messageId(int messageId){
        message.putInt(InboundMessage.MESSAGE_ID_POS,messageId);
    }
    public void connectionId(int connectionId){
        message.putInt(InboundMessage.CONNECTION_ID_POS,connectionId);
    }
    public void sessionId(int sessionId){
        message.putInt(InboundMessage.SESSION_ID_POD,sessionId);
    }
    public void sequence(int sequence){
        message.putInt(InboundMessage.SEQUENCE_POS,sequence);
    }
    public void timestamp(long timestamp){
        message.putLong(InboundMessage.TIMESTAMP_POS,timestamp);
    }
    public void payload(byte[] payload){
        message.position(InboundMessage.PAYLOAD_POS);
        message.put(payload);
        payloadSize = payload.length;
    }
    public byte[] message(){
        byte[] payload = new byte[InboundMessage.PAYLOAD_POS+payloadSize];
        message.position(InboundMessage.ACK_POS);
        message.get(payload);
        return payload;
    }
}
